taichi/backends/vulkan/codegen_vulkan.cpp

```diff
@@ -257,208 +312,365 @@ class TaskCodegen : public IRVisitor {
   void visit(GlobalStoreStmt *stmt) override {
     TI_ASSERT(stmt->width() == 1);
     const auto dt = stmt->val->element_type();
-    emit("{} = {};", at_buffer(stmt->dest, dt),
-         store_as_int_bits(stmt->val->raw_name(), dt));
+    spirv::Value buffer_ptr; 
+    spirv::Value val = ir_->query_value(stmt->val->raw_name());
+    if (ptr_to_buffers_.at(stmt->dest) == BuffersEnum::Root) {
+      buffer_ptr = ir_->query_value(stmt->dest->raw_name());
+      buffer_ptr.flag = spirv::ValueKind::kVariablePtr; // make this value could store/load
+    } else {
+      buffer_ptr = at_buffer(stmt->dest, dt);
+    }
+    ir_->store_variable(buffer_ptr, ir_->make_value(spv::OpBitcast, ir_->i32_type(), val));
   }
 
   void visit(GlobalLoadStmt *stmt) override {
     TI_ASSERT(stmt->width() == 1);
     auto dt = stmt->element_type();
-    const auto loaded_int = at_buffer(stmt->src, dt);
-    emit("const {} {} = {};", opengl_data_type_name(dt), stmt->raw_name(),
-         load_from_int_bits(loaded_int, dt));
+    spirv::Value buffer_ptr;
+    spirv::Value val;
+    if (ptr_to_buffers_.at(stmt->src) == BuffersEnum::Root) {
+      buffer_ptr = ir_->query_value(stmt->src->raw_name());
+      buffer_ptr.flag = spirv::ValueKind::kVariablePtr; // make this value could store/load
+    } else {
+      buffer_ptr = at_buffer(stmt->src, dt);
+    }
+    val = ir_->make_value(spv::OpBitcast, ir_->get_primitive_type(dt),
+          ir_->load_variable(buffer_ptr, ir_->i32_type()));
+    ir_->register_value(stmt->raw_name(), val);
```

Comment:
If I'm understanding this correctly (loading a i32 value and cast it to the target primitive type), avoid using BitCast. GPU raw loads from structured buffers have built-in type conversion, so it is better to have multiple typed buffers binding to the same location (like the current GL backend)

Reply:
This is to make the atomics work for floating point numbers. Since we have generated a struct for each SNode (instead of manually computing the position in the global buffer), not using the bitcast means that we would have to re-generate the OpAccessChain twice, once for the target type and the other for i32.

That said, in the fture we can add another impl  that follows the GL's approach to see how it performs.

Reply:
I think OpAccessChain may not actually compile to an instruction, same with bit cast. However it is more likely that bitcast is compiled into separate instruction instead of using the load store unit's free modifier.